File: Flutter Money card/lib/models/product.dart
/// Product entity conforming to M0 V10 specification.
/// Note: Product.tags was explicitly removed in M0 V10; category is a multi-select String array.
class Product {
  final String id;
  final String branchId;
  final String itemName;
  final List<String> category;
  final double price;
  final String status; // 'ACTIVE' | 'INACTIVE'
  final String? createdAt;
  final String? updatedAt;

  const Product({
    required this.id,
    required this.branchId,
    required this.itemName,
    this.category = const [],
    required this.price,
    this.status = 'ACTIVE',
    this.createdAt,
    this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    List<String> parsedCategories = [];
    if (json['category'] is List) {
      parsedCategories = (json['category'] as List<dynamic>)
          .map((e) => e.toString())
          .toList();
    } else if (json['categories'] is List) {
      parsedCategories = (json['categories'] as List<dynamic>)
          .map((e) => e.toString())
          .toList();
    } else if (json['category'] is String) {
      parsedCategories = [json['category'] as String];
    }

    return Product(
      id: json['id'] as String? ?? '',
      branchId: json['branchId'] as String? ?? json['branch_id'] as String? ?? '',
      itemName: json['itemName'] as String? ?? json['item_name'] as String? ?? '',
      category: parsedCategories,
      price: (json['price'] as num?)?.toDouble() ?? 0.0,
      status: json['status'] as String? ?? 'ACTIVE',
      createdAt: json['createdAt'] as String? ?? json['created_at'] as String?,
      updatedAt: json['updatedAt'] as String? ?? json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'branchId': branchId,
        'itemName': itemName,
        'category': category,
        'price': price,
        'status': status,
        if (createdAt != null) 'createdAt': createdAt,
        if (updatedAt != null) 'updatedAt': updatedAt,
      };
}

/// Inventory item conforming to M0 V10 specification.
class InventoryItem {
  final String id;
  final String branchId;
  final String productId;
  final int quantity;
  final String? updatedAt;

  const InventoryItem({
    required this.id,
    required this.branchId,
    required this.productId,
    required this.quantity,
    this.updatedAt,
  });

  factory InventoryItem.fromJson(Map<String, dynamic> json) {
    return InventoryItem(
      id: json['id'] as String? ?? '',
      branchId: json['branchId'] as String? ?? json['branch_id'] as String? ?? '',
      productId: json['productId'] as String? ?? json['product_id'] as String? ?? '',
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      updatedAt: json['updatedAt'] as String? ?? json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'branchId': branchId,
        'productId': productId,
        'quantity': quantity,
        if (updatedAt != null) 'updatedAt': updatedAt,
      };
}

/// Combined product and its active branch inventory quantity
class ProductWithInventory {
  final Product product;
  final int quantity;

  const ProductWithInventory({
    required this.product,
    required this.quantity,
  });

  factory ProductWithInventory.fromJson(Map<String, dynamic> json) {
    return ProductWithInventory(
      product: Product.fromJson(json),
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    final map = product.toJson();
    map['quantity'] = quantity;
    return map;
  }
}

File: Frontend Money Card/src/types/product.ts
// ─── Product & Inventory Entities (M0 V10 Section 3, 4 & 19) ──────
// Products and inventory are branch-scoped.
// V10: Category is multi-select string array. Tags field is removed.

export interface Product {
  id: string;
  branchId: string;
  itemName: string;
  category: string[];
  price: number;
  status: 'ACTIVE' | 'INACTIVE';
  createdAt: string;
  updatedAt: string;
}

export interface InventoryItem {
  id: string;
  branchId: string;
  productId: string;
  quantity: number;
  updatedAt: string;
}

export interface ProductWithInventory extends Product {
  quantity: number;
  inventoryUpdatedAt?: string;
  branchName?: string;
}

// ─── CSV Import Specification (M0 V10 Section 12 & 19) ─────────
// Required columns: itemName, category, price. Tags is removed.

export interface CsvRowInput {
  rowNumber: number;
  itemName: string;
  category: string[];
  price: number;
}

export interface CsvValidationError {
  rowNumber: number;
  field: string;
  reason: string;
}

export interface CsvImportPreview {
  branchId: string;
  totalRows: number;
  validRows: CsvRowInput[];
  invalidRows: CsvValidationError[];
  createsCount: number;
  updatesCount: number;
  previewToken: string;
}

export interface CreateProductRequest {
  branchId: string;
  itemName: string;
  category: string[];
  price: number;
  status?: 'ACTIVE' | 'INACTIVE';
  initialQuantity?: number;
}

export interface UpdateProductRequest {
  itemName?: string;
  category?: string[];
  price?: number;
  status?: 'ACTIVE' | 'INACTIVE';
}

export interface UpdateInventoryRequest {
  quantity: number;
}
